File: Sources/macOS/Sources/OnScreen/Models/PetEntity.swift
import Combine
import Schwifty
import SwiftUI
import Yage

class PetEntity: Entity {
    @Inject private var settings: AppConfig
    
    private var disposables = Set<AnyCancellable>()

    public init(of species: Species, in world: World) {
        super.init(
            species: species,
            id: PetEntity.id(for: species),
            frame: PetEntity.initialFrame(for: species),
            in: world
        )
        resetSpeed()
        setInitialPosition()
        setInitialDirection()
        bindGravity()
        bindBounceOffPets()
    }
    
    private func bindBounceOffPets() {
        settings.$bounceOffPetsEnabled
            .sink { [weak self] in self?.setBounceOffPets(enabled: $0) }
            .store(in: &disposables)
    }
    
    private func setBounceOffPets(enabled: Bool) {
        guard let bounce = capability(for: BounceOnLateralCollisions.self) else { return }
        if enabled {
            bounce.customCollisionsFilter = { _ in true }
        } else {
            bounce.customCollisionsFilter = nil
        }
    }
    
    private func bindGravity() {
        settings.$gravityEnabled
            .sink { [weak self] in self?.setGravity(enabled: $0) }
            .store(in: &disposables)
    }

    override open func set(state: EntityState) {
        super.set(state: state)
        if case .move = state { resetSpeed() }
    }

    public func resetSpeed() {
        speed = PetEntity.speed(
            for: species,
            size: frame.width,
            settings: settings.speedMultiplier
        )
    }
    
    func setInitialPosition() {
        let randomX = worldBounds.width * .random(in: 0.05...0.95)
        let randomY: CGFloat

        if capability(for: WallCrawler.self) != nil {
            randomY = worldBounds.height - frame.height
        } else {
            randomY = 60 
        }
        frame.origin = CGPoint(x: randomX, y: randomY)
    }

    func setInitialDirection() {
        direction = .init(dx: 1, dy: 0)
    }
    
    override open func kill() {
        disposables.removeAll()
        super.kill()
    }
}

// MARK: - Incremental Id

extension PetEntity {
    static func id(for species: Species) -> String {
        nextNumber += 1
        return "\(species.id)-\(nextNumber)"
    }

    private static var nextNumber = 0
}

// MARK: - Speed

extension PetEntity {
    internal static let baseSpeed: CGFloat = 30

    static func initialFrame(for species: Species) -> CGRect {
        @Inject var appConfig: AppConfig
        return CGRect(square: appConfig.petSize * species.scale)
    }
    
    static func speed(for species: Species, size: CGFloat, settings: CGFloat) -> CGFloat {
        species.speed * speedMultiplier(for: size) * settings
    }

    static func speedMultiplier(for size: CGFloat) -> CGFloat {
        let sizeRatio = size / PetSize.defaultSize
        return baseSpeed * sizeRatio
    }
}

// MARK: - Pet Size

public struct PetSize {
    public static let defaultSize: CGFloat = DeviceRequirement.iPhone.isSatisfied ? 50 : 75
    public static let minSize: CGFloat = 25
    public static let maxSize: CGFloat = 350
}

File: Sources/windows/onscreen/models/pet_entity.py
import random
from config.config import Config
from di.di import Dependencies
from onscreen.models.pet_size import PetSize
from yage.capabilities import Gravity
from yage.models.entity import Entity
from yage.models.entity_state import EntityState
from yage.models.species import Species
from yage.models.world import World
from yage.utils.geometry import Point, Rect, Vector


class PetEntity(Entity):
    def __init__(self, species: Species, world: World):
        self.config = Dependencies.instance(Config)
        pet_size = self.config.pet_size.value
        gravity_enabled = self.config.gravity_enabled.value

        super().__init__(
            species,
            PetEntity.next_id(species),
            Rect(0, 0, pet_size, pet_size),
            world
        )
        self.reset_speed()
        self.place_in_random_position()
        self._set_initial_direction()
        self.set_gravity_enabled(gravity_enabled)

    def reset_speed(self):
        self.speed = PetEntity.default_speed(
            self.species,
            self.frame.width,
            self.config.speed_multiplier.value
        )

    def place_in_random_position(self):
        random_x = random.uniform(0.1, 0.75) * self.world_bounds.width
        self.frame.origin = Point(random_x, 5)

    def _set_initial_direction(self):
        self.direction = Vector(1, 0)

    def set_gravity_enabled(self, enabled: bool):
        gravity = self.capability(Gravity)
        if not gravity:
            return
        gravity.is_enabled = enabled
        if not enabled:
            if self.direction.dy > 0:
                self.direction = Vector(1, 0)
            self.state = EntityState.MOVE

    @classmethod
    def initial_frame(cls) -> Rect:
        config = Dependencies.instance(Config)
        return Rect(0, 0, config.pet_cize, config.pet_cize)

    @classmethod
    def default_speed(cls, species: Species, size: float, settings: float) -> float:
        return species.speed * cls.speed_multiplier(size) * settings

    @classmethod
    def speed_multiplier(cls, size: float) -> float:
        size_ratio = size / PetSize.default_size
        return 30 * size_ratio

    @classmethod
    def next_id(cls, species):
        if not hasattr(cls, 'incremental_id'):
            cls.incremental_id = 0
        cls.incremental_id += 1
        return f'{species.id}-{cls.incremental_id}'
